### PR TITLE
Implement Backend Fastapi limiter

### PR DESCRIPTION
**Pull request by kiritowu**
Second PR for #175 .

Changes:
- Include [fastapi-limiter](https://github.com/long2ice/fastapi-limiter) into backend requirements.txt
- Created `UserRateLimiter` and `APIClientRateLimiter` as fastapi dependencies for API rate limiting with proper error handling code  (e.g. `OasstErrorCode.TOO_MANY_REQUESTS`)
- Update `docker-compose.yml to start redis server on backend-dev and establish proper connection in docker compose environment
- Add rate-limiter dependecies to `/api/v1/task` endpoint as PoC.

To-Do:
- Identify list of endpoints that requires API rate limiting and verify its compatibility with existing rate limiter
- Perform load testing to study the appropriate parameters for both user and api-client of each endpoints.

### CODE REVIEW

File: backend/oasst_backend/api/deps.py [hunk 1]
```diff
@@ -84,3 +85,60 @@ def get_trusted_api_client(
...
+
+        super().__init__(times, milliseconds, seconds, minutes, hours, identifier)
+
+    async def __call__(self, request: Request, response: Response, api_key: str = Depends(get_api_key)) -> None:
+        # Skip if rate limiting is disabled
+        if not settings.RATE_LIMIT:
+            return
+
+        # Attempt to retrieve api_key and user information
+        api_key = request.headers.get("X-API-Key") or request.query_params.get("api_key")
```
Comment (yk) [inline backend/oasst_backend/api/deps.py:108]: aren't we getting the api_key via injection?

File: backend/oasst_backend/api/deps.py [hunk 2]
```diff
@@ -84,3 +85,60 @@ def get_trusted_api_client(
...
```
Comment (yk) [inline backend/oasst_backend/api/deps.py:136]: same as above

**Review by yk — APPROVED** [review]
LGTM, thank you. I just left small questions

File: backend/oasst_backend/api/deps.py (hunk 1 above)
Comment (kiritowu) [reply backend/oasst_backend/api/deps.py:108]: Oops that was a mistake, let me quickly make a fix.

**Comment by yk** [comment]
thank you!

**Pull request merged**